core/iwasm/common/wasm_shared_memory.c

```diff
@@ -489,17 +497,27 @@ wasm_runtime_atomic_notify(WASMModuleInstanceCommon *module, void *address,
     WASMModuleInstance *module_inst = (WASMModuleInstance *)module;
     uint32 notify_result;
     AtomicWaitInfo *wait_info;
+    WASMSharedMemNode *node;
+    bool out_of_bounds;
 
     bh_assert(module->module_type == Wasm_Module_Bytecode
               || module->module_type == Wasm_Module_AoT);
 
-    if ((uint8 *)address < module_inst->memories[0]->memory_data
-        || (uint8 *)address + 4 > module_inst->memories[0]->memory_data_end) {
+    node = search_module((WASMModuleCommon *)module_inst->module);
+    os_mutex_lock(&node->shared_mem_lock);
+    out_of_bounds =
+        ((uint8 *)address < module_inst->memories[0]->memory_data
+         || (uint8 *)address + 4 > module_inst->memories[0]->memory_data_end);
+    os_mutex_unlock(&node->shared_mem_lock);
+
+    if (out_of_bounds) {
         wasm_runtime_set_exception(module, "out of bounds memory access");
         return -1;
     }
 
+    os_mutex_lock(&node->shared_mem_lock);
     wait_info = acquire_wait_info(address, false);
+    os_mutex_unlock(&node->shared_mem_lock);
```

Comment:
Same as above:
```C
    if (node)
        os_mutex_lock(&node->shared_mem_lock);
    wait_info = acquire_wait_info(address, false);
    if (node)
        os_mutex_unlock(&node->shared_mem_lock);
```
I have a test, after adding checks, the spec cases can run success.

Reply:
And maybe here we can avoid locking twice:
```C
    node = search_module((WASMModuleCommon *)module_inst->module);
    if (node)
        os_mutex_lock(&node->shared_mem_lock);
    out_of_bounds =
        ((uint8 *)address < module_inst->memories[0]->memory_data
         || (uint8 *)address + 4 > module_inst->memories[0]->memory_data_end);

    if (out_of_bounds) {
        wasm_runtime_set_exception(module, "out of bounds memory access");
        if (node)
            os_mutex_unlock(&node->shared_mem_lock);
        return -1;
    }

    wait_info = acquire_wait_info(address, false);
    if (node)
        os_mutex_unlock(&node->shared_mem_lock);
```

Reply:
Not sure what's failing now, the error is different; I can reproduce it locally, but I see the same error even if I run the spec tests from the main branch

Reply:
Missed adding unlock before return -1, seems that causes the spec test failure.

Reply:
thanks, that worked!